src/Native/Runtime/amd64/StubDispatch.S

```diff
@@ -0,0 +1,171 @@
+//
+DEFINE_INTERFACE_DISPATCH_STUB 1
+DEFINE_INTERFACE_DISPATCH_STUB 2
+DEFINE_INTERFACE_DISPATCH_STUB 4
+DEFINE_INTERFACE_DISPATCH_STUB 8
+DEFINE_INTERFACE_DISPATCH_STUB 16
+DEFINE_INTERFACE_DISPATCH_STUB 32
+DEFINE_INTERFACE_DISPATCH_STUB 64
+
+// Initial dispatch on an interface when we don't have a cache yet.
+LEAF_ENTRY RhpInitialInterfaceDispatch, _TEXT
+ALTERNATE_ENTRY RhpInitialDynamicInterfaceDispatch
+
+        // Just tail call to the cache miss helper.
+        jmp     C_FUNC(RhpInterfaceDispatchSlow)
+
+LEAF_END RhpInitialInterfaceDispatch, _TEXT
+
+// Cache miss case, call the runtime to resolve the target and update the cache.
+NESTED_ENTRY RhpInterfaceDispatchSlow, _TEXT, NoHandler
+
+        .equ RIDS_ReservedStack, 30h + 50h + 80h + 8h     // Scratch space, transition frame, xmm registers and padding
+
+        alloc_stack         RIDS_ReservedStack
+
+        // Preserve the argument registers in the scratch space across the helper call. Note that we depend on these
+        // registers (which may contain GC references) being spilled before we build the PInvokeTransitionFrame below
+        // due to the way we build a stack range to report to the GC conservatively during a collection.
+        mov     [rsp + 8*0], rdi
+        mov     [rsp + 8*1], rsi
+        mov     [rsp + 8*2], rdx
+        mov     [rsp + 8*3], rcx
+        mov     [rsp + 8*4], r8
+        mov     [rsp + 8*5], r9
+
+        save_xmm128_postrsp xmm0, (30h + 50h + 16*0)
+        save_xmm128_postrsp xmm1, (30h + 50h + 16*1)
+        save_xmm128_postrsp xmm2, (30h + 50h + 16*2)
+        save_xmm128_postrsp xmm3, (30h + 50h + 16*3)
+        save_xmm128_postrsp xmm4, (30h + 50h + 16*4)
+        save_xmm128_postrsp xmm5, (30h + 50h + 16*5)
+        save_xmm128_postrsp xmm6, (30h + 50h + 16*6)
+        save_xmm128_postrsp xmm7, (30h + 50h + 16*7)
+        END_PROLOGUE
+
+        // Build PInvokeTransitionFrame. This is only required if we end up resolving the interface method via
+        // a callout to a managed ICastable method. In that instance we need to be able to cope with garbage
+        // collections which in turn need to be able to walk the stack from the ICastable method, skip the
+        // unmanaged runtime portions and resume walking at our caller. This frame provides both the means to
+        // unwind to that caller and a place to spill callee saved registers in case they contain GC
+        // references from the caller.
+
+        // Save caller's rip.
+        mov     rax, [rsp + RIDS_ReservedStack]
+        mov     [rsp + 30h + 8*0], rax
+
+        // Save caller's rbp.
+        mov     [rsp + 30h + 8*1], rbp
+
+        // Zero out the Thread*, it's not used by the stackwalker.
+        xor     rax, rax
+        mov     [rsp + 30h + 8*2], rax
+
+        // Set the flags.
+        mov     dword ptr [rsp + 30h + 8*3], PTFF_SAVE_ALL_PRESERVED + PTFF_SAVE_RSP
+
+        // Save callee saved registers.
+        mov     [rsp + 30h + 8*4], rbx
```

Comment:
Like for the argument registers, it seems using PUSH would have better performance.